### PR TITLE
RegistrationServiceImpl and RegistrationServiceImplTest were implemented

### PR DESCRIPTION
**Pull request by MykolaVoronov**

### CODE REVIEW

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 1]
```diff
@@ -1,11 +1,40 @@
...
+    private void validateUserData(User user) {
+        if (user == null
+                || user.getLogin() == null
+                || user.getAge() == null
+                || user.getPassword() == null
+                || user.getLogin().length() < MINIMUM_LOGIN_LENGTH
+                || user.getAge() < MINIMUM_USER_AGE
+                || user.getPassword().length() < MINIMUM_PASSWORD_LENGTH) {
+            throw new RuntimeException("Invalid user data!");
+        }
```
Comment (ShurionUA) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:30]: I think we should throw narrow Exception for better understanding - Invalid login or Invalid age or Invalid password

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 2]
```diff
@@ -1,11 +1,40 @@
...
+    }
+
+    private void checkIsLoginAlreadyRegistered(String login) {
+        for (User currentUser : Storage.people) {
+            if (currentUser.getLogin().equals(login)) {
+                throw new RuntimeException("User with this email address already exists!");
```
Comment (ShurionUA) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:36]: login != email) Login can be different to email.

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 3]
```diff
@@ -1,11 +1,40 @@
...
 
+import core.basesyntax.dao.StorageDao;
+import core.basesyntax.dao.StorageDaoImpl;
+import core.basesyntax.db.Storage;
 import core.basesyntax.model.User;
 
 public class RegistrationServiceImpl implements RegistrationService {
+    private static final int MINIMUM_USER_AGE = 18;
+    private static final int MINIMUM_PASSWORD_LENGTH = 6;
+    private static final int MINIMUM_LOGIN_LENGTH = 1;
```
Comment (ShurionUA) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:11]: We don't have any criteria for login's length.

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        user.setLogin("");
+        assertThrows(RuntimeException.class, () -> registrationService.register(user));
+    }
+
+    @Test
+    void register_nullLogin_notOk() {
+        user.setLogin(null);
+        assertThrows(RuntimeException.class, () -> registrationService.register(user));
+    }
+}
```
Comment (ShurionUA) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:84]: I think we should check if there are really 2 or more different users registered in database.

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java (hunk 1 above)
Comment (MykolaVoronov) [reply src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:30]: The task says that in case of invalid data, you should throw a RuntimeException. But it is not said that you need to throw for each exception separately

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java (hunk 3 above)
Comment (MykolaVoronov) [reply src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:11]: Yes, but for the logical operation of the program, I consider it necessary

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        user.setPassword(null);
+        assertThrows(RuntimeException.class, () -> registrationService.register(user));
+    }
+
+    @Test
+    void register_nullLogin_notOk() {
+        user.setLogin(null);
+        assertThrows(RuntimeException.class, () -> registrationService.register(user));
+    }
+}
```
Comment (ShurionUA) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:78]: We need to test for empty login

**Review by ShurionUA — APPROVED** [review]

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+import core.basesyntax.model.User;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+class RegistrationServiceImplTest {
+
```
Comment (Nick97-git) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:13]: empty line)

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    void setUp() {
+        Storage.people.clear();
+        user.setAge(18);
+        user.setLogin("qwerty@gmail.com");
+        user.setPassword("qwerty123");
+    }
+
+    @Test
+    void register_existedUser_notOk() {
+        Storage.people.add(user);
```
Comment (Nick97-git) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:33]: let's just register the same user twice and we won't need this line

**Review by Nick97-git — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 8]
```diff
@@ -1,11 +1,33 @@
...
 public class RegistrationServiceImpl implements RegistrationService {
+    private static final int MINIMUM_USER_AGE = 18;
+    private static final int MINIMUM_PASSWORD_LENGTH = 6;
+    private StorageDao storageDao = new StorageDaoImpl();
 
     @Override
     public User register(User user) {
-        return null;
+        validateUserData(user);
+        if (Storage.people.contains(user)) {
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:16]: look at the `Storage` methods, there is the method that you need for this check, also, add this check into if of `validateUserData`

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        user.setLogin("");
+        assertThrows(RuntimeException.class, () -> registrationService.register(user));
+    }
+
+    @Test
+    void register_nullLogin_notOk() {
+        user.setLogin(null);
+        assertThrows(RuntimeException.class, () -> registrationService.register(user));
+    }
+}
```
Comment (Nick97-git) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:83]: add empty line)

**Review by Nick97-git — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 10]
```diff
@@ -1,11 +1,30 @@
...
     @Override
     public User register(User user) {
-        return null;
+        validateUserData(user);
+        return storageDao.add(user);
+    }
+
+    private void validateUserData(User user) {
+        if (user == null
+                || storageDao.get(user.getLogin()) != null
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:20]: this check must be the last one

**Review by Nick97-git — CHANGES_REQUESTED** [review]

**Review by Nick97-git — APPROVED** [review]
